### PR TITLE
zhuanzhuan's zset implement and other fix

### PR DESCRIPTION
**Pull request by piaoairy219**
1 implements zset's 6 commands: zadd, zrange, zrevrange, zrem, zcard, zscore, and delete, expire, gc also can clear relative data
2 the list-zip-threshold is config item
3 slove the problem when connection is closed,  pipeline will miss un-handled commands, by pipe and add SOCKET_CLOSE_FLAG
4 slove flushdb and flushall problem by delete 50000 keys once in many times
5 gc seek prefix once, delete 10 keys prefixed, this will make gc quick up
6 fix problem that gc leader changes frequently in many nodes
7 add cost trace in steps of AutoCommit

### CODE REVIEW

**Comment by shafreeck** [comment]
This is a large PR with 52 commits and it has many conflictions with the master branch, we should review the PR carefully and schedule a plan for it.

File: command/common.go [hunk 1]
```diff
@@ -55,7 +55,6 @@ func (t *Base) UnmarshalBinary(data []byte) error {
 	return nil
 }
 
-//Verify token auth
 func Verify(token, key []byte) ([]byte, error) {
```
Comment (shafreeck) [inline command/common.go]: Export methods should have a comment

File: encoding/resp/resp.go [hunk 2]
```diff
@@ -119,6 +122,22 @@ func (r *Encoder) Array(size int) error {
 	return err
 }
 
+func (r *Encoder) BulkStrings(strs [][]byte) error {
```
Comment (shafreeck) [inline encoding/resp/resp.go]: BulkStrings can be combined with BulkString and Array. RESP should implement the primitive data types only.

File: README.md [hunk 3]
```diff
@@ -1,20 +1,4 @@
-# Titan
```
Comment (shafreeck) [inline README.md]: These should not be deleted

File: client.go [hunk 4]
```diff
@@ -6,39 +6,64 @@ import (
...
-	conn   net.Conn
-	exec   *command.Executor
-	r      *bufio.Reader
+	cliCtx  *context.ClientContext
+	server  *Server
+	conn    net.Conn
+	exec    *command.Executor
+	r       *bufio.Reader
+	
+	eofLock sync.Mutex	//the lock of reading_writing 'eof'
```
Comment (shafreeck) [inline client.go]: The problem of pipeline commands has been fixed.

File: client.go [hunk 5]
```diff
@@ -112,8 +147,9 @@ func (c *client) serve(conn net.Conn) error {
 				zap.String("traceid", ctx.TraceID),
 				zap.String("command", ctx.Name))
 		}
+		
 		c.exec.Execute(ctx)
-		cancel()
+		cancel()
```
Comment (shafreeck) [inline client.go]: Format your code with gofmt

File: command/command.go [hunk 6]
```diff
@@ -61,17 +62,30 @@ func Integer(w io.Writer, v int64) OnCommit {
...
 			resp.ReplyBulkString(w, string(a[i]))
+			if i % 10 == 9 {
+				zap.L().Debug("reply 10 bulk string", zap.Int64("cost(us)", time.Since(start).Nanoseconds()/1000))
+				start = time.Now()
+			}
 		}
 	}
 }
 
+func BytesArrayOnce(w io.Writer, a [][]byte) OnCommit {
```
Comment (shafreeck) [inline command/command.go]: What's the difference with BytesArray?

File: command/command.go [hunk 7]
```diff
@@ -146,7 +160,21 @@ func Call(ctx *Context) {
 
 	feedMonitors(ctx)
 	start := time.Now()
-	cmdInfoCommand.Proc(ctx)
+	//wangzongsheng add below
+	if ctx.Name == "flushdb" || ctx.Name == "flushall" {
```
Comment (shafreeck) [inline command/command.go]: flushdb / flushall has been implemented by a more performant way

File: command/connection.go [hunk 8]
```diff
@@ -20,13 +20,15 @@ func Auth(ctx *Context) {
...
+		ctx.Client.Authenticated = false
+		return
 	}
 	ctx.Client.Authenticated = true
 	metrics.GetMetrics().ConnectionOnlineGaugeVec.WithLabelValues(ctx.Client.Namespace).Dec()
 	metrics.GetMetrics().ConnectionOnlineGaugeVec.WithLabelValues(string(namespace)).Inc()
 	ctx.Client.Namespace = string(namespace)
 	ctx.Client.DB.Namespace = string(namespace)
-	ctx.Client.Authenticated = true
+	//ctx.Client.Authenticated = true  //wangzongsheng note this, it's repeat
```
Comment (shafreeck) [inline command/connection.go]: It's OK to remove this line. It was a bug.

File: command/connection.go [hunk 9]
```diff
@@ -20,13 +20,15 @@ func Auth(ctx *Context) {
 	namespace, err := Verify(token, serverauth)
 	if err != nil {
 		resp.ReplyError(ctx.Out, "ERR invalid password")
+		ctx.Client.Authenticated = false
```
Comment (shafreeck) [inline command/connection.go]: ctx.Client.Authenticated  is already false, not necessary to assign again.

File: command/init.go [hunk 10]
```diff
@@ -79,6 +79,14 @@ func init() {
 		// sets
 		"sadd":     SAdd,
 		"smembers": SMembers,
+
+		//zsets(sorted sets)
+		"zadd":         ZAdd,
```
Comment (shafreeck) [inline command/init.go]: Format these lines with gofmt

File: command/lists.go [hunk 11]
```diff
@@ -8,18 +8,19 @@ import (
 	"github.com/meitu/titan/db"
 )
 
-var (
-	// ListZipThreshold indicates to create a ziplist when it is exceeded to push elements
-	ListZipThreshold = 100
-)
+//wangzongsheng note this, we use config item
```
Comment (shafreeck) [inline command/lists.go]: It's OK to remove the outdated code.

**Comment by piaoairy219** [comment]
have created new one(delete zhuanzhuan's flushDB/flushAll/quick gc fix) see https://github.com/distributedio/titan/pull/77

**Pull request closed without merging**